File: leetcode/dynamic_programming/combination_sum4/python/solution.py
from typing import List


class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [0] * (target + 1)
        dp[0] = 1
        for i in range(1, target + 1):
            for num in nums:
                if i - num >= 0:
                    dp[i] += dp[i - num]
        return dp[target]


def solve(nums: List[int], target: int) -> int:
    return Solution().combinationSum4(nums, target)


if __name__ == "__main__":
    # quick sanity-run
    print(solve([1, 2, 3], 4))

File: leetcode/dynamic_programming/combination_sum4/go/solution.go
package main

import "fmt"

func Solve(nums []int, target int) int {
	dp := make([]int, target+1)
	dp[0] = 1
	for i := 1; i <= target; i++ {
		for _, num := range nums {
			if i-num >= 0 {
				dp[i] += dp[i-num]
			}
		}
	}
	return dp[target]
}

func main() {
	fmt.Println(Solve([]int{1, 2, 3}, 4))
}
